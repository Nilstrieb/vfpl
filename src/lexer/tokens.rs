use crate::error::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenType,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    Please,

    Ident(String),
    // Keywords
    As,
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    Initialize,
    Variable,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    Out,
    This,
    Create,
    Function,
    Parameter,
    Parameters,
    That,
    Returns,
    Call,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Does,
    Has,
    Is,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,
}

impl Display for TokenType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenType::Please => f.write_str("keyword `please`"),
            TokenType::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenType::As => f.write_str("keyword `as`"),
            TokenType::Add => f.write_str("keyword `add`"),
            TokenType::Subtract => f.write_str("keyword `subtract`"),
            TokenType::Multiply => f.write_str("keyword `multiply`"),
            TokenType::Divide => f.write_str("keyword `divide`"),
            TokenType::Modulo => f.write_str("keyword `modulo`"),
            TokenType::Initialize => f.write_str("keyword `initialize`"),
            TokenType::Variable => f.write_str("keyword `variable`"),
            TokenType::With => f.write_str("keyword `with`"),
            TokenType::The => f.write_str("keyword `the`"),
            TokenType::Value => f.write_str("keyword `value`"),
            TokenType::Of => f.write_str("keyword `of`"),
            TokenType::Set => f.write_str("keyword `set`"),
            TokenType::To => f.write_str("keyword `to`"),
            TokenType::From => f.write_str("keyword `from`"),
            TokenType::By => f.write_str("keyword `by`"),
            TokenType::Take => f.write_str("keyword `take`"),
            TokenType::End => f.write_str("keyword `end`"),
            TokenType::Check => f.write_str("keyword `check`"),
            TokenType::Whether => f.write_str("keyword `whether`"),
            TokenType::Then => f.write_str("keyword `then`"),
            TokenType::Do => f.write_str("keyword `do`"),
            TokenType::Otherwise => f.write_str("keyword `otherwise`"),
            TokenType::Break => f.write_str("keyword `break`"),
            TokenType::Out => f.write_str("keyword `out`"),
            TokenType::This => f.write_str("keyword `this`"),
            TokenType::Create => f.write_str("keyword `create`"),
            TokenType::Function => f.write_str("keyword `function`"),
            TokenType::Parameter => f.write_str("keyword `parameter`"),
            TokenType::Parameters => f.write_str("keyword `parameters`"),
            TokenType::That => f.write_str("keyword `that`"),
            TokenType::Returns => f.write_str("keyword `returns`"),
            TokenType::Go => f.write_str("keyword `go`"),
            TokenType::Sleep => f.write_str("keyword `sleep`"),
            TokenType::Absent => f.write_str("keyword `absent`"),
            TokenType::Null => f.write_str("keyword `null`"),
            TokenType::NoValue => f.write_str("keyword `no value` )"),
            TokenType::Undefined => f.write_str("keyword `undefined`"),
            TokenType::True => f.write_str("keyword `true`"),
            TokenType::False => f.write_str("keyword `false`"),
            TokenType::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenType::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Dot => f.write_str("`.`"),
            TokenType::ParenOpen => f.write_str("`(`"),
            TokenType::ParenClose => f.write_str("`)`"),
            TokenType::Comma => f.write_str("`,`"),
        }
    }
}
