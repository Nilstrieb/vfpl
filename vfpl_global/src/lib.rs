mod global_ctx;
mod session;

pub use global_ctx::GlobalCtx;
pub use session::Session;

/// Only we need to depend on `lasso`, use re-exports for things downstream crates might need
pub use lasso::Spur;
