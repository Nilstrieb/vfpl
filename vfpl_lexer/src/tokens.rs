use std::fmt::{Display, Formatter};
use vfpl_error::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    CondKeyword(CondKeyword),
    Ident(String),
    // Keywords
    Please,
    As,
    Initialize,
    Variable,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    This,
    Create,
    Function,
    Call,
    And,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Not,
    Or,
    Repeat,
    Return,
    While,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,

    /// The last token
    Eof,
}

/// A conditional keyword that can be used as an identifier
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CondKeyword {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    Out,
    Parameter,
    Parameters,
    That,
    Returns,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Does,
    Has,
    Is,
    Have,
    Greater,
    Less,
    Than,
    Equal,
}

impl CondKeyword {
    fn as_str(&self) -> &'static str {
        match self {
            CondKeyword::Add => "add",
            CondKeyword::Sub => "sub",
            CondKeyword::Mul => "mul",
            CondKeyword::Div => "div",
            CondKeyword::Mod => "mod",
            CondKeyword::With => "with",
            CondKeyword::The => "the",
            CondKeyword::Value => "value",
            CondKeyword::Of => "of",
            CondKeyword::Set => "set",
            CondKeyword::To => "to",
            CondKeyword::From => "from",
            CondKeyword::By => "by",
            CondKeyword::Take => "take",
            CondKeyword::Out => "out",
            CondKeyword::Parameter => "parameter",
            CondKeyword::Parameters => "parameters",
            CondKeyword::That => "that",
            CondKeyword::Returns => "returns",
            CondKeyword::No => "no",
            CondKeyword::Argument => "argument",
            CondKeyword::Arguments => "arguments",
            CondKeyword::Go => "go",
            CondKeyword::Sleep => "sleep",
            CondKeyword::Does => "does",
            CondKeyword::Has => "has",
            CondKeyword::Is => "is",
            CondKeyword::Have => "have",
            CondKeyword::Greater => "greater",
            CondKeyword::Less => "less",
            CondKeyword::Than => "than",
            CondKeyword::Equal => "equal",
        }
    }
}

impl Display for TokenKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenKind::Please => f.write_str("keyword `please`"),
            TokenKind::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenKind::As => f.write_str("keyword `as`"),
            TokenKind::Initialize => f.write_str("keyword `initialize`"),
            TokenKind::Variable => f.write_str("keyword `variable`"),
            TokenKind::End => f.write_str("keyword `end`"),
            TokenKind::Check => f.write_str("keyword `check`"),
            TokenKind::Whether => f.write_str("keyword `whether`"),
            TokenKind::Then => f.write_str("keyword `then`"),
            TokenKind::Do => f.write_str("keyword `do`"),
            TokenKind::Otherwise => f.write_str("keyword `otherwise`"),
            TokenKind::Break => f.write_str("keyword `break`"),
            TokenKind::This => f.write_str("keyword `this`"),
            TokenKind::Create => f.write_str("keyword `create`"),
            TokenKind::Function => f.write_str("keyword `function`"),
            TokenKind::Absent => f.write_str("keyword `absent`"),
            TokenKind::Null => f.write_str("keyword `null`"),
            TokenKind::NoValue => f.write_str("keyword `novalue` )"),
            TokenKind::Undefined => f.write_str("keyword `undefined`"),
            TokenKind::True => f.write_str("keyword `true`"),
            TokenKind::False => f.write_str("keyword `false`"),
            TokenKind::Not => f.write_str("keyword `not`"),
            TokenKind::Or => f.write_str("keyword `or`"),
            TokenKind::Call => f.write_str("keyword `call`"),
            TokenKind::Repeat => f.write_str("keyword `repeat`"),
            TokenKind::Return => f.write_str("keyword `return`"),
            TokenKind::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenKind::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenKind::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenKind::Dot => f.write_str("`.`"),
            TokenKind::ParenOpen => f.write_str("`(`"),
            TokenKind::ParenClose => f.write_str("`)`"),
            TokenKind::Comma => f.write_str("`,`"),
            TokenKind::And => f.write_str("keyword `and`"),
            TokenKind::While => f.write_str("keyword `while`"),
            TokenKind::Eof => f.write_str("end of file"),
            TokenKind::CondKeyword(cond) => f.write_str(&format!("keyword `{}`", cond.as_str())),
        }
    }
}

impl Token {
    /// Token with a Span of one character
    pub fn new_from_single(kind: TokenKind, idx: usize) -> Self {
        Token {
            kind,
            span: Span::single(idx),
        }
    }

    pub fn new(kind: TokenKind, start: usize, end: usize) -> Self {
        Token {
            kind,
            span: Span::start_end(start, end),
        }
    }

    pub fn new_from_len(kind: TokenKind, start: usize, end: usize) -> Self {
        Token {
            kind,
            span: Span::start_len(start, end),
        }
    }

    pub fn eof() -> Self {
        Token {
            span: Span::eof(),
            kind: TokenKind::Eof,
        }
    }
}
