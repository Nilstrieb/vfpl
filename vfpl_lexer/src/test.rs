use crate::tokens::Token;
use crate::{Lexer, LexerResult};
use vfpl_global::GlobalCtx;

fn lex(code: &str) -> LexerResult<Vec<Token>> {
    Lexer::new(&code.to_lowercase(), GlobalCtx::test_ctx()).compute_tokens()
}

#[test]
fn one_number() {
    insta::assert_debug_snapshot!(lex("1"));
}

#[test]
fn leading_zeros() {
    insta::assert_debug_snapshot!(lex("000012"));
}

#[test]
fn minus_number() {
    insta::assert_debug_snapshot!(lex("-12"));
}

#[test]
fn minus_with_leading_zeros() {
    insta::assert_debug_snapshot!(lex("-0000012"));
}

#[test]
fn alphabetical_char_in_number() {
    insta::assert_debug_snapshot!(lex("12a"));
}

#[test]
fn decimal() {
    insta::assert_debug_snapshot!(lex("13.35253"));
}

#[test]
fn decimal_with_leading_zeros() {
    insta::assert_debug_snapshot!(lex("000001.15"));
}

#[test]
fn decimal_with_minus_and_leading_zeros() {
    insta::assert_debug_snapshot!(lex("-00000012.23"));
}

#[test]
fn decimal_with_alphabetical_char() {
    insta::assert_debug_snapshot!(lex("00001.as244"));
}

#[test]
fn please_keyword() {
    insta::assert_debug_snapshot!(lex("please"));
}

#[test]
fn string() {
    insta::assert_debug_snapshot!(lex("\"Hello all\""));
}

#[test]
fn string_with_escape_chars() {
    insta::assert_debug_snapshot!(lex(r#""What is a \"unit test\"?""#));
}

#[test]
fn string_no_end() {
    insta::assert_debug_snapshot!(lex("\"Hahaha this is going to destroy it"));
}

#[test]
fn should_be_ident() {
    insta::assert_debug_snapshot!(lex("please hugo"));
}

#[test]
fn all_keywords() {
    let tokens = [
        "as",
        "add",
        "sub",
        "mul",
        "div",
        "modulo",
        "initialize",
        "variable",
        "with",
        "the",
        "value",
        "of",
        "set",
        "to",
        "from",
        "by",
        "take",
        "end",
        "check",
        "whether",
        "then",
        "do",
        "otherwise",
        "break",
        "out",
        "this",
        "create",
        "function",
        "parameter",
        "parameters",
        "that",
        "returns",
        "call",
        "no",
        "argument",
        "arguments",
        "and",
        "go",
        "sleep",
        "absent",
        "null",
        "noValue",
        "undefined",
        "true",
        "false",
        "does",
        "has",
        "is",
        "not",
        "have",
        "greater",
        "less",
        "than",
        "equal",
        "or",
        "repeat",
        "return",
        "while",
        "structure",
        "field",
        "fields",
        "define",
    ]
    .map(lex);
    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn normal_line_of_code() {
    insta::assert_debug_snapshot!(lex(
        "please initialize variable x as Integer with the value of 15.\n"
    ));
}

#[test]
fn normal_line_of_code_decimal() {
    insta::assert_debug_snapshot!(lex(
        "please initialize variable x as Integer with the value of 15.3.\n"
    ));
}
