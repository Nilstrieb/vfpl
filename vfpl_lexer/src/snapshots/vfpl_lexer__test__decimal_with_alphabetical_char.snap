---
source: vfpl_lexer/src/test.rs
expression: "lex(\"00001.as244\")"

---
Err(
    InvalidCharacter(
        Span {
            start: 6,
            end: 7,
        },
    ),
)
